Make Content Recipe Step Idempotent

deanmarcussen opened the pull request:
Fixes https://github.com/OrchardCMS/OrchardCore/issues/5230 
Fixes https://github.com/OrchardCMS/OrchardCore/issues/3089
Fixes https://github.com/OrchardCMS/OrchardCore/issues/4694

Putting this forward as a draft to demonstrate the issue that occurs when using deployment plans to manage content from a stage to prod environment.

The problem that occurs then, when importing a new version of an existing content item, is that the old content remains listed as published or latest, because the old content is not queried and updated.

~~I've done all this in the recipe step, without adding an import handler to the content manager, as if we did so it should be designed.~~

~~This makes recipe imports idempotent, but does not include any verb patching as has been suggested in https://github.com/OrchardCMS/OrchardCore/issues/1970. This is because deployment sources don't currently include any option to mutate a content item, they only care about published items.~~

~~technically a recipe could be modified by hand to try and patch an item, but right now that wouldn't work, and is not currently designed to do so, they are currently designed to operate solely on whether the contentitemversion is recognized or not.~~

~~So this also makes some arbitrary decisions about what content should remain published.
e.g. if it's in the recipe and a new version id, it takes priority over any existing drafts, and becomes published.~~

~~Otherwise the possibility exists that an item is drafted on a prod server, a new version published on the stage server, and if the prod draft item remained `latest` then it would still potentially be behind what was imported from the stage server. But we can design that better if we want to be less arbitrary.~~

~~One of the issues that remains is potential for problems because this is not done in a `contentmanager.import` (but I understand why that was removed after all the O1 issues around it) is issue https://github.com/OrchardCMS/OrchardCore/issues/3405.
The potential issues exists that if a scoped index provider tries to retrieve an imported version from the content manager it could end up with a new entity/query, because of the `ContentManagerSession` cache.~~

~~So we may want to design an `Import` method, that covers that better, than using the existing `Create` method~~

deanmarcussen commented on src/OrchardCore.Modules/OrchardCore.Contents/Recipes/ContentStep.cs:
```diff
@@ -36,24 +37,52 @@ public async Task ExecuteAsync(RecipeExecutionContext context)
-                    // Replaces the id to force the current item to be updated
-                    existing.Id = contentItem.Id;
-                    _session.Save(existing);
+
```
removed this as it makes no sense to save an item that has just been retrieved, and not mutated.
If we did want to do mutatations, this would be the place.
Removing this saves around 10 sql calls to indexes etc, to update / delete / create indexes, that have not changed.

Skrypt replied:
You sure because _session.Save is a no op method?

Skrypt replied:
Ok here what it does is that it will save the existing content item if nothing is done afterward. So if it doesn't find any relatedContentItems it will still call a _session.Save(existing)

So we have

contentItem = passed from the step context
existing = retrieved from DB and specific version (so could be draft or published)
relatedContentItem = any contentItem (drafted or published) related with the contentItem passed from the step context.

Then you iterate over these relatedContentItem to save each of them and you set these as Published or Draft based on contentItem.Published or ContentItem.Latest.

Skrypt replied:
Ok we can remove because it will still be called (I assume)

Skrypt commented on src/OrchardCore.Modules/OrchardCore.Contents/Recipes/ContentStep.cs:
```diff
@@ -36,24 +37,52 @@ public async Task ExecuteAsync(RecipeExecutionContext context)
+
+                    // Reset published and latest to imported values as CreateAsync sets these values arbitrarily.
+                    contentItem.Published = isPublished;
+                    contentItem.Latest = isLatest;
```
Here I'm not sure why you need to reset these values as you don't do a CreateAsync in this "else" statement.

deanmarcussen replied:
so actually the code shown here is hard to read.

I have removed the else statement.

As it does / did nothing except force an existing item that had just been retrieved from the database (and not mutated) to be resaved.

I'm not saying we should necessarily remove it permanently, but unless we do some mutation on existing content (i.e. the same version id) then it makes no sense.

I'm not sure that mutating an existing contentitemversion is the right thing to do either. 
The idea should be (IMHO) that altering a contentitem via a recipe, means creating a new version to do so, i.e. that's what a deployment plan should do, update content that has changed.

Unless we want to support editing a recipe by hand, to mutate existing content without creating a new version id.

Skrypt replied:
I agree

Skrypt commented on src/OrchardCore.Modules/OrchardCore.Contents/Recipes/ContentStep.cs:
```diff
@@ -36,24 +37,52 @@ public async Task ExecuteAsync(RecipeExecutionContext context)
+                            continue;
+                        }
+
+                        if (isPublished)
```
Why not use contentItem.Published directly?

deanmarcussen replied:
Because depending on what is done in the `CreateAsync` or handlers, the value can be mutated. 
But here we should to force it to what is supplied in the recipe, so that the result is always the same (and is a defined, understandable result, that is idempotent, because we define what should happen going in, and always get the same output)

Skrypt replied:
Ok I agree too

deanmarcussen commented:
Little bit of further commentary on this.

as it stands the recipe step will also fail to run both the `UpdateHandlers` and the `PublishedHandlers`

This will potentially cause problems for `Autoroute`, if routing url's are changed in the imported item.

My gut feeling is we could design an `Import` method on the `ContentManager` which could run all the required handlers, and place items in the correct `ContentManagerSession` cache, rather than trying to use the existing methods

Skrypt commented:
While I was working on the ApiController and added the using the proper sequence of methods called from the ContentManager I tested to push a content item with an AutoroutePart and a DisplayText. The result was that if we provide an AutoroutePart.Path in the ContentItem it will take precendence over the DisplayText and use it instead. Though, this way it won't valide if the AutoroutePart.Path passed is going to be a duplicate. We assume that if you pass an AutoroutePart.Path directly that this is because you know that it won't generate an Autoroute for you and won't check for an existing same route.

So when here we use the ContentManager in the ContentSteps we only call CreateAsync and never UpdateAsync which is wrong. It should do a UpdateAndCreateAsync() at least to call the proper events. Then in each of the services that should not execute while we are running a recipe then we should have a Flag that tells us if it should be omitted to execute it in the context of a recipe execution. Some more explanation made in https://github.com/OrchardCMS/OrchardCore/issues/2609

Skrypt commented:
It might be an option to never use the ContentManager in a IRecipeStepHandler. The only thing we would require is to create recipe steps for resetting/rebuilding a Lucene index after importing data. Also, if a content item is added to the ContentItemIndex table it should also be added to the IndexingTask table. Right now, the CreateIndexingTaskContentHandler is calling the IIndexingTaskManager to create these Indexing tasks. This is why normally we should require to use the ContentManager for doing these operations. I think I need to evaluate both scenarios. Else, the only thing missing in the ContentStep would be to call this IndexingTaskManager.

Skrypt commented:
If we are doing this on Latest or Published content items then in AllContentDeploymentSource.cs we should also change it so that it exports Published or Latest content items.

deanmarcussen commented:
My preference would to be even more arbitrary.

Remove the published and latest properties from the recipe schema, and when importing an item is always published and latest.

Why you ask? 

Because importing an item that is not latest and not published means the item will just be lost in the database, so makes no sense.

Skrypt commented:
Ok the idea is to migrate data wathever it is because I could import and old recipe file as far as I know and a content item could have changed on the deployed host too. So it would make any more sense if we took a published/latest only record from a previous deployment and tried to update the database with it even if the current content item has been versioned more recently. 

So I think here whe should allow importing drafts and history too because that's what is a synchonization of data between 2 tenants is. Even deleted records are only soft deleted because we want to keep an history of what happened and eventually have an audit trail that shows who did what and maybe even be able to comment on each audit. 

But here, if we only export the published/latest records we assume that you have already the same data in both database minus some changes made on the staging tenant. Though this is a bad assumption because while you are creating content items on your staging website anyone can also iterate on the prod website.

Skrypt commented:
Please fix also https://github.com/OrchardCMS/OrchardCore/issues/2609 as part of this PR.
I'm off for the next 2 weeks working on my own stuff.
Pretty simple. remove the ContentManager, don't use it "remove the CreateAsync" and use the IndexingTaskManager directly in ContentStep.cs.

Skrypt commented:
Appveyor acting funny... not even triggering a build.

agriffard commented:
@deanmarcussen Can you please fix the PR?

Skrypt commented on src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs:
```diff
@@ -403,6 +403,142 @@ public async Task CreateAsync(ContentItem contentItem, VersionOptions options)
+            {
+                existingItem = await GetVersionAsync(contentItem.ContentItemVersionId);
+            }
+
```
We could also return if the content item already exist to make sure we don't import these.

```
            if (!String.IsNullOrEmpty(contentItem.ContentItemId))
            {
                if(await GetAsync(contentItem.ContentItemId) != null)
                {
                    return;
                }
            }
```

Skrypt replied:
Forget this comment. Actually doing this would return on every items.

Skrypt commented:
I'm going to test it right now and make some manual tests on this.

deanmarcussen commented:
> I'm going to test it right now and make some manual tests on this.

Cool cheers :) wip. and I have broken a GraphQL test.

Reasonably well tested is the recipe import when you have a different version id.
Not so well tested is the code for when you have the same version id, but are trying to modify the content... Does work, just a few permutations to test.

TODO: events for throttling background tasks (lucene).
Not sure if I should add these events at start of recipe import, or start of the content step.

Events will be something like `RecipeRunning` `RecipeCompleted` and try catch for failures.
Think maybe better to add at Recipe start level in case there is some use for other background tasks unrelated to Lucene... thoughts?

deanmarcussen commented on test/OrchardCore.Tests/Apis/ContentManagement/DeploymentPlans/BlogPostCreateDeploymentPlanTests.cs:
```diff
@@ -0,0 +1,152 @@
+                        Assert.Equal("new version", newVersion?.DisplayText);
+                        Assert.True(newVersion?.Latest);
+                        Assert.True(newVersion?.Published);
+                    });
```
Would appreciate some input from anyone on how I am building these integration tests up.

I've co-opted @Jetski5822 GraphQL API tests, and introduced access to the DI container of the built tenant.
This is used to configure remote recipe deployment (as we have no api for doing that - it could be done via a custom recipe, but...

For the general Act of the test I am using the API's or controller endpoints.

For testing it however, I'm resolving the `YesSql.ISession` and querying the database directly, as we have no API endpoint that would allow me to test it otherwise.

I think it's proving useful, as it's identified a number of issues, particularly in the `ContentApiController` where records were being incorrectly stored, and sometimes duplicated (fixed)

But opinions of others here would be helpful if they think it is an acceptable integration testing pattern

jtkech replied:
Funny, we communicate through the API with a tenant on which we can create a scope ;)

I'm okay with this pattern where we build a shell scope but only to check the results.

Skrypt commented on src/OrchardCore.Modules/OrchardCore.Autoroute/Handlers/AutoroutePartHandler.cs:
```diff
@@ -112,6 +112,13 @@ public override async Task UpdatedAsync(UpdateContentContext context, AutorouteP
             // Compute the Path only if it's empty
             if (!String.IsNullOrWhiteSpace(part.Path))
```
Remove this because it prevents from having a unique path for each of them.

deanmarcussen replied:
I think this is ok because otherwise it would try and generate the path every time, which would not work if the path has been set manually, and not with a pattern.

I have not yet added a `slugify` check here like we talked about, but I would rather check with `slugify` as well, and then make a unique path based on the data provided, than generating with a guid. (unless you have reason that is a bad idea?). That way the path (or alias) provided, is as close to what was requested by either the api or content item generated from code as possible.

There is another option.
We talked about validation.
We could make the ContentApiController run the display manager on the content item, and return any validation errors without creating the content item.

That might not be a bad idea.

I saw for example that it is possible to make a `BlogPost` without all the required fields (taxonomy and tags) so while the `BlogPost` creates, it is not a valid content item

Skrypt replied:
Ok here you need to take in consideration that from an API Call the event will be triggered. If you pass in the part.Path a string for example "Foo Bar". This string will fail to generate a unique path from that first "if" statement.

Skrypt commented on src/OrchardCore.Modules/OrchardCore.Alias/Handlers/AliasPartHandler.cs:
```diff
@@ -40,6 +40,13 @@ public async override Task UpdatedAsync(UpdateContentContext context, AliasPart
             // Compute the Alias only if it's empty
             if (!String.IsNullOrEmpty(part.Alias))
```
Remove

Skrypt commented on src/OrchardCore.Modules/OrchardCore.Autoroute/Handlers/AutoroutePartHandler.cs:
```diff
@@ -112,6 +112,13 @@ public override async Task UpdatedAsync(UpdateContentContext context, AutorouteP
+                // Validate path is unique here as api services do not provide validation.
+                if (!await IsPathUniqueAsync(part.Path, part))
+                {
+                    part.Path = await GenerateUniquePathAsync(part.Path, part);
```
In this method if the string passed is empty generate a path with a guid

Skrypt commented on src/OrchardCore.Modules/OrchardCore.Alias/Handlers/AliasPartHandler.cs:
```diff
@@ -40,6 +40,13 @@ public async override Task UpdatedAsync(UpdateContentContext context, AliasPart
+                // Validate path is unique here as api services do not provide validation.
+                if (!await IsAliasUniqueAsync(part.Alias, part))
+                {
+                    part.Alias = await GenerateUniqueAliasAsync(part.Alias, part);
```
Generate path with string empty as a guid

deanmarcussen commented:
Fixes https://github.com/OrchardCMS/OrchardCore/issues/1970
Fixes https://github.com/OrchardCMS/OrchardCore/issues/5731
Fixes https://github.com/OrchardCMS/OrchardCore/issues/2609

Ok, so this is probably ready to review now.

I've introduced two new methods in the content manager
- CreateContentItemVersionAsync
- UpdateContentItemVersionAsync

One will create a new content item version for an existing content item, and process all the necessary events, and remove existing drafts etc

The other will mutate an existing content item, and validate existing drafts or published items, depending on the values provided by the mutated items.

The big difference between the existing UpdateAsync and CreateAsync is that both of these methods are interested in previous published / draft versions.

I'm claiming idempotency for the content recipe step as it will no longer forget to remove previous drafts, and can be run multiple times and produce the same desired outcome very time.

I've also fixed and tested some issues with autoroute unicity and the `.Merge` method

There is a test proving that lucene indexes are now rebuilt on recipe step execution, and I've made some comments on that issue https://github.com/OrchardCMS/OrchardCore/issues/2609 as to why I think we don't need any events, and why I think it's actually pretty good at the moment.

Please don't merge until @sebastienros has had a chance to review

jtkech commented:
@deanmarcussen @Skrypt 

Yes, easy to say but just some thoughts after a quick look (would need more time) on the alias part.

- Because of the alias auto generation we do some validation stuff into the handler. Before we were not doing that if the alias is already set to prevent some duplicate stuff, assuming that the validation has already been done by the driver or a previous handler execution.

- But this is not the case when triggered by an API / recipe / workflow, so now we do the validation even if the alias is already set, but this not prevent anymore some duplicate stuff.

- **So, maybe good to find a way to use the same validation code from everywhere**, by the driver and the handler itself. Maybe a separate validation service per part (maybe implemented by the driver itself), and maybe **a contextual flag indicating that the part has already been validated**.

I will try something this night if i have time.

sebastienros commented:
To do some validation outside of the editor we need an Importing/Imported event in the content handler. This way we can update the values on api calls. The only other option, which would also make things idempotent, would be to have a Validate event, that would fail if there is a conflict with an existing item. Because change the route during import makes things not idempotent.

deanmarcussen commented:
@jtkech did you have time to try anything?

I maybe trying to do too much with this pr at one time.

The validation issue exists, and has done for sometime, so perhaps is better fixed with a good design on another pr.

That said, I lean towards @sebastienros suggestion of a `Validate` event, and failing either the import, api call, or workflow create. 
This might be more ubiquitous than an importing/imported event, because it could be more useful across the various ways recipes, apis, or workflow, (or code) that can be use to create and import. 
For example the recipe import is not the same as the Api Post. 

The Api Post builds a new version, and mutates that, whereas the recipe is intended to import a version that exists in a file.

I suspect it would be better for an import / Api call to fail with errors, than to mutate the content item (as I have done here, and may revert).

Also noticed, for example, in the Api Post for a BlogPost, if the required fields are not set, they are not validated, and yet the content item is still created.

I'm not sure that I would suggest moving all the UI validation to a validate event, but if we did have one, that was at least used for the `Alias` and `Autoroute` parts where unicity is important, then we would also have a mechanism for api users to implement their own validation requirements? 

Thoughts?

sebastienros commented:
This `Validate` method would not be called on UI event (UpdatingAsync). Because UI validation needs more context, like the html field that is wrong, and might also change the values before re-editing. The Validate method should not change the content item. It would still return messages, in a `ContentValidateResults` like the `IdentityResult` does. It can't use an `out` argument as it needs to be async.

jtkech commented:
@deanmarcussen sorry i didn't try anything ;)

I understand what you said and what @sebastienros is suggesting, but i would need to focus on it more. I'll try to think about it this night.

deanmarcussen commented:
No problem @jtkech I also understand what @sebastienros has suggested with `ContentValidateResults` and it sounds like a good plan. I will work on it in the next couple of days

deanmarcussen commented on src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs:
```diff
@@ -403,6 +405,124 @@ public async Task CreateAsync(ContentItem contentItem, VersionOptions options)
+                var possibleEvictionVersions = await _session.Query<ContentItem, ContentItemIndex>()
+                    .Where(x => x.ContentItemId.IsIn(possibleEvictionIds) && (x.Latest || x.Published))
+                    .ListAsync();
+
```
Refactored this to make two queries
- queries for existing content item versions
- queries for the other latest and / or published versions of those items that need evicting 

Batched it to 500 content items per batch of queries.

deanmarcussen commented on src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs:
```diff
@@ -403,6 +405,124 @@ public async Task CreateAsync(ContentItem contentItem, VersionOptions options)
+        {
+            return UpdateContentItemVersionAsync(updatedVersion, updatedVersion);
+        }
+
```
I've left these two methods in, as it was suggested on the issue that we should have them.
When called in this manner (i.e. not importing) they would make sql queries to discover if other version need evicting.

They could be removed if `ImportAsync()` is enough

deanmarcussen commented on src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs:
```diff
@@ -403,6 +405,124 @@ public async Task CreateAsync(ContentItem contentItem, VersionOptions options)
+                        // So it is important to only import changed items.
+                        var jImporting = JObject.FromObject(importingItem);
+                        var jOriginal = JObject.FromObject(originalVersion);
+
```
This comparison is important, because the deployment step will export all published and latest content, even if it has not changed (how would it know that detail...)

So this will no-op if the `ContentItem` has not changed.

deanmarcussen commented on src/OrchardCore.Modules/OrchardCore.Alias/Drivers/AliasPartDisplayDriver.cs:
```diff
@@ -55,12 +57,12 @@ private async Task ValidateAsync(AliasPart alias, IUpdateModel updater)
             if (alias.Alias?.Length > MaxAliasLength)
             {
-                updater.ModelState.AddModelError(Prefix, nameof(alias.Alias), S["Your alias is too long. The alias can only be up to {0} characters.", MaxAliasLength]);
+                updater.ModelState.AddModelError(Prefix, nameof(alias.Alias), S[DefaultMaxAliasLengthError, MaxAliasLength]);
```
Can't use constants here or the PO localizer won't extract them. So keep the values the same, but as strings in the Validate Drivers

jtkech commented on src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs:
```diff
@@ -403,6 +405,124 @@ public async Task CreateAsync(ContentItem contentItem, VersionOptions options)
+
+            var batchedContentItems = contentItems.Skip(skip).Take(take);
+
+            while (batchedContentItems.Any())
```
Not sure but just because i thought about it. Maybe good to create a shell scope per batch. Unless we want a complete successful import or nothing

jtkech commented on src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs:
```diff
@@ -493,5 +613,259 @@ public async Task<ContentItem> CloneAsync(ContentItem contentItem)
+            }
+
+            // Re-enlist content item here in case of session queries.
+            _session.Save(contentItem);
```
I'm afraid that when importing, e.g if more than one batch, in the next loop another session query will trigger an auto flush and then clear the session saved list. Or as suggested above, and if possible, create a scope per batch.

Maybe we need to do the save through another service that does a session save immediately and then do it again at the end of the shell scope, as i did in my branch for committing though an `IDocumentStore`, see #6008

deanmarcussen replied:
@jtkech I will have to check this further, but has some behaviour in YesSql changed around `AutoFlush()` ?

I see a lot of comments about re-enlisting the save in our code, but when I look in the YesSql code, `FlushAsync` always Saves the entities, before clearing the updated / saved lists? 
But it does not commit the transaction, until it is disposed

My testing seems to confirm this, but as I say, there are a lot of comments about re-enlisting for save in our code.

The only reason I use the batching here, is for the version queries, to avoid this issue https://github.com/sebastienros/yessql/issues/187

So yes, if we need to we could a session shell scope, though I was preferring to be able to do it all as one session, to be able to cancel the entire import, if one of the validations fail. 

I will adapt one of the tests to run multiple batches (but not commit it, as it would slow the CI) to check the behaviour here

jtkech commented:
@deanmarcussen i did a little review

Would need more time and fork your branch to see it more in details, will do as i have time

deanmarcussen commented:
Note to self: update the new `UpdateContentItem` Workflow Task to use validate methods

jtkech commented:
@deanmarcussen 

> I see a lot of comments about re-enlisting the save in our code, but when I look in the YesSql code, FlushAsync always Saves the entities, before clearing the updated / saved lists?

Sorry i was wrong, this is no more an issue precisely because of the new auto flush, but there is still a potential issue if after a session save, **we do a query and then another mutation** (no matter where it is done), another session save is needed to take into account the last mutation.

So here maybe still an issue e.g if an handler does a query and then also does a mutation. That's why in my last comment in #6008 i talk about a more global solution where the session saves would need to be done as soon as possible, but this through another service, `ISessionHelper` or `IDocumentStore` in the future, that will recall the session saves before commiting the session.

This to ensure that any handler can retrieve the current item being mutated, while ensuring that any mutated item will be persisted. Maybe i will open an issue and / or a PR.

> So yes, if we need to we could a session shell scope, though I was preferring to be able to do it all as one session, to be able to cancel the entire import, if one of the validations fail.

Okay make sense.

@deanmarcussen, because i also saw it in #5979, why are you using `OrdinalIgnoreCase` to compare unique identifiers, why not just using `id1 == id2`?

deanmarcussen commented:
> we do a query and then another mutation

This is making good sense, so yes

> @deanmarcussen, because i also saw it in #5979, why are you using `OrdinalIgnoreCase` to compare unique identifiers, why not just using `id1 == id2`?

I have been using `OrdinalIgnoreCase` on content item ids because every time I put in a pr without it @sebastienros reviews and says use `OrdinalIgnoreCase` 

Most recent review : https://github.com/OrchardCMS/OrchardCore/pull/5653

Would be good to clarify that if you think it is an issue (I can see the issue, but is our convention `OrdinalIgnoreCase` or not on these unique identifiers?)

jtkech commented:
> Most recent review : #5653

He didn't say to use OrdinalIgnoreCase, he just asked you if better ;)

> Would be good to clarify that if you think it is an issue (I can see the issue, but is our convention OrdinalIgnoreCase or not on these unique identifiers?)

Maybe not an issue as there are so many possible values, but yes interesting to see why comparing an unique id to 2^25 different values.

deanmarcussen commented on src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs:
```diff
@@ -403,6 +405,124 @@ public async Task CreateAsync(ContentItem contentItem, VersionOptions options)
+
+        public Task<ContentValidateResult> UpdateContentItemVersionAsync(ContentItem updatingVersion, ContentItem updatedVersion)
+        {
+            return UpdateContentItemVersionAsync(updatedVersion, updatedVersion);
```
This was the intention @jtkech 
`UpdateContentItemVersionAsync(updatingVersion, updatedVersion, null)`

Good spot, updated.

deanmarcussen commented on src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs:
```diff
@@ -403,6 +405,124 @@ public async Task CreateAsync(ContentItem contentItem, VersionOptions options)
+
+                var evictionVersions = await _session.Query<ContentItem, ContentItemIndex>()
+                    .Where(x => x.ContentItemId.IsIn(evictionIds) && (x.Latest || x.Published))
+                    .ListAsync();
```
The reason for the two queries here, is that the first is against a `ContentItemVersionId`, to get the existing versions, that are going to be updated.

Then the second is against a `ContentItemId` to get the versions that are in the database that are latest and/or published.

So I could not see a clear way to get both of what we need from a single query.

Because a `ContentItemVersion` maybe being updated, but two other versions of the `ContentItemId` may also be published and/or latest, so they are the ones we are looking for in the second query.

deanmarcussen commented on src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs:
```diff
@@ -403,6 +405,124 @@ public async Task CreateAsync(ContentItem contentItem, VersionOptions options)
+
+                foreach (var version in evictionVersions)
+                {
+                    await LoadAsync(version);
```
Yes, done knowing `LoadAsync()` will no-op if already called

deanmarcussen commented:
> He didn't say to use OrdinalIgnoreCase, he just asked you if better ;)

This is true :)

> Maybe not an issue as there are so many possible values, but yes interesting to see why comparing an unique id to 2^25 different values.

It is tricky, but for me, I am ok with `OrdinalIgnoreCase` on what we are generating from a guid, i.e. the place for conflicts in doing `OrdinalIgnoreCase` would not be from a generated id, but from someone changing an id manually, and upper casing it to make it "different". 

But for me `4z0gkzcqrqf6rzybn43ya1pxxv` is ok to consider the same as `4z0gkzcqrqf6rzybn43ya1pXXV`

Am I wrong, or missing something?

I am happy with the shape of this pr, but have a few more tests to run, and things to check over.

If you had any time to review further that would be great.

I have noted some of the places where the potential `_sessionHelper.Save()` should be used, as that looks like a good pr to resolve any flushing / mutated post flush issues.

(hoping that might be merged before, so I can implement it here, and test drive it)

deanmarcussen commented:
/cc @duncanhoggan I am tweaking your new `UpdateContent` activity to include limited validation  (Autoroute, and Alias uniqueness)

deanmarcussen commented on src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs:
```diff
@@ -403,6 +405,124 @@ public async Task CreateAsync(ContentItem contentItem, VersionOptions options)
+                    ContentItem originalVersion = null;
+                    if (!String.IsNullOrEmpty(importingItem.ContentItemVersionId))
+                    {
+                        originalVersion = versions.FirstOrDefault(x => String.Equals(x.ContentItemVersionId, importingItem.ContentItemVersionId, StringComparison.OrdinalIgnoreCase));
```
Ah, interesting, I did not know YesSql was not case sensitive there. 

So I learn something to ;)

I wonder if it is dialect related, because @jptissot had an issue somewhere, about postgresql being case sensitive on the `like` query syntax

deanmarcussen commented on src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs:
```diff
@@ -493,5 +653,261 @@ public async Task<ContentItem> CloneAsync(ContentItem contentItem)
+        private async Task RemoveVersionsAsync(ContentItem contentItem, IEnumerable<ContentItem> evictionVersions)
+        {
+            IEnumerable<ContentItem> activeVersions;
+            if (evictionVersions != null)
```
yes, should be `if (evictionVersions == null)` (so done)

It is possible evictionVersions is null (intentionally), because this will get called by either
- ImportAsync (which will do the larger query across a number of items for the evictionVersions), and will not be null, or
- `UpdateContentItemVersionAsync(), - the standalone public api method, which will then call the private helpers with a null eviction versions.

So done this way to reduce the requirement for (multiple) queries and keep the logic in one place.

deanmarcussen commented on src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs:
```diff
@@ -493,5 +653,261 @@ public async Task<ContentItem> CloneAsync(ContentItem contentItem)
+                contentItem.ContentItemVersionId = _idGenerator.GenerateUniqueId(contentItem);
+            }
+
+            // Remove previous published or latest items or they will continue to be listed as published or latest.
```
Done

deanmarcussen commented on src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs:
```diff
@@ -493,5 +653,261 @@ public async Task<ContentItem> CloneAsync(ContentItem contentItem)
+                // the imported (which we assume is the version that wins) content.
+                await RemoveVersionsAsync(contentItem, evictionVersions);
+            }
+            // When neither imported or latest the operation will create a database record
```
Done

deanmarcussen commented on src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs:
```diff
@@ -403,6 +406,147 @@ public async Task CreateAsync(ContentItem contentItem, VersionOptions options)
+
+                // Query for versions that may be evicted.
+                var publishedLatestVersions = versions.Where(x => x.Latest && x.Published);
+                var reducedContentItems = batchedContentItems.Except(publishedLatestVersions);
```
> `batchedContentItems.Except(publishedLatestVersions)` will not work 

Think you are right. 

Will investigate this more, and see if that suggested query will do the trick

deanmarcussen replied:
I like it @jtkech single query working. cheers 👍

deanmarcussen commented:
@Skrypt can you complete your review on this?

I reverted the change to including latest in the all content deployment source - was felt that should be a setting, which we could do at anytime.

Not quite ready to merge yet, would wait for @jtkech to have more time, and I have some other testing to complete as well, but I am happy with the shape of it

Skrypt commented:
Will do.

jtkech commented:
> Not quite ready to merge yet, would wait for @jtkech to have more time, and I have some other testing to complete as well, but I am happy with the shape of it

Will do too, will need to focus again on session helper first

jtkech commented:
@deanmarcussen

I did a new PR #6054 related to the `UpdateContentTask` and i saw here the new `ValidateAsync()` that you use in it. For infos here i now get the item from the content manager with `DraftRequired`, most of the time, if directly triggered by a content event it retrieves the item being mutated, otherwise it retrieves it from the store and a new draft version may be created if the item is `Versionable`.

That said, the problem is that when calling before in this activity `UpdateAsync()` a session save is done, so it's too late to validate it as in a WF we can't do a session cancel. Maybe we would need a session cancel activity but this may introduce other issues. So maybe here we would have to 1st work on a cloned item in memory for the validation.

I will think about it tomorrow.

deanmarcussen commented:
> @deanmarcussen

> That said, the problem is that when calling before in this activity `UpdateAsync()` a session save is done, so it's too late to validate it as in a WF we can't do a session cancel. Maybe we would need a session cancel activity but this may introduce other issues. So maybe here we would have to 1st work on a cloned item in memory for the validation.
> 
> I will think about it tomorrow.

Not sure if this is an issue or not @jtkech 

In the `ValidateAsync()` method it calls a `_session.Cancel()` if the validation fails.

I tested this in the unit tests with an item when that item has already had save called on it (and a query run causing an `AutoFlush()`.

The sql to update the saved item runs, and then the transaction is aborted by the cancel, so there is no actual update to the database.

I would have to check again to be 100%, but the tests confirm the behaviour, although I may have misunderstood what you are saying

Though what we might now need is some `Workflow` validation handlers

jtkech commented:
> In the ValidateAsync() method it calls a _session.Cancel() if the validation fails

Cool then, i thought it had to be done by the caller

> Though what we might now need is some Workflow validation handlers

Not sure i understand here, can you give me more infos

Did you try to run 2 imports concurently ;)

deanmarcussen commented:
@jtkech 

> > In the ValidateAsync() method it calls a _session.Cancel() if the validation fails
> 
> Cool then, i thought it had to be done by the caller
> 
> > Though what we might now need is some Workflow validation handlers

```
            var validateContext = new ValidateContentContext(contentItem);

            await Handlers.InvokeAsync((handler, context) => handler.ValidatingAsync(context), validateContext, _logger);

            await ReversedHandlers.InvokeAsync((handler, context) => handler.ValidatedAsync(context), validateContext, _logger);

            if (!validateContext.ContentValidateResult.Succeeded)
            {
                _session.Cancel();
            }
```

 For infos @jtkech all the validation is done in the `ValidatingAsync` handler, and there is the opportunity for a custom handler to override a failure (and force it to succeed) in `ValidatedAsync`, but if the validation has failed after that handler the session would be cancelled.

So at some point we could also have a workflow event for the `ValidatingAsync` and `ValidatedAsync` events, however it's not necessary right now. 

The current implementation is only to secure unicity for `Autoroute` and `Alias`, but it would allow users to implement their own custom validation logic for API / Workflow related creates, as they can't rely on the drivers to validate required components.

> Did you try to run 2 imports concurently ;)

Haha I shall, I suspect it will not fail, but it will be last one wins, if both imports are working on the same content item.

deanmarcussen commented:
@jtkech 

> Haha I shall, I suspect it will not fail, but it will be last one wins, if both imports are working on the same content item.

Yes, works concurrently (ran some tests twice with a short `Task.Delay`) in between. Second import wins, but doesn't crash, nor create unexpected entries.

I will mark this ready, and if you have some time for a deeper look, that would be great.

We could wait for the `session.helper()` or if you are happy to approve, we could merge it and implement the `session.helper()` as and when.

For me I have tested enough now I think, but your second look would be great

jtkech commented:
@deanmarcussen thanks for the infos

> So at some point we could also have a workflow event for the ValidatingAsync and ValidatedAsync events, however it's not necessary right now.

**Okay i got it** as we already have `ContentCreatedEvent`, `ContentUpdatedEvent` and so on.

But as i commented in some related issues / prs, when used inline just after a `ContentEvent`, **maybe we need some dedicated activities acting as handlers**. Handlers are referenced by the content manager and then can't reference it, **so "handler" activities should not use content manager methods** as `UpdateAsync()` calling other handlers, not have the responsability to do `PublishAsync()`, not call `ValidateAsync()` that may do a session cancel (i'll talk about cancel below).

When not triggered by a content event that's okay, e.g. `UpdateContentTask` can acts as a mini controller, otherwise we would need e.g. a `ContentUpdatedHandlerTask` acting as an handler in place of trying to use the 1st for all cases.

**But just some thinkings that should not block this PR** but that we will have to consider.

> We could wait for the session.helper()

What i already saw is that if we do a save, then a query + mutation e.g. in an handler, we then need to do another save to take into account the last mutation, **so here okay would be good to use the session helper** to re-do it at the end of the scope just before commiting.

But i also saw that if we do a mutation + save, then if we do a session cancel but then another query + save the mutation is persisted. So it is important that a cancel is the last operation before committing, as in controllers. **So here the session helper re-doing the save is not good**, we would need to do all cancel through the session helper so that it could check it before re-doing a save.

A commit release the transaction + connection, a cancel release the transaction, both don't release the session so the identityMap is not cleared, okay to retrieve an entity but here a new transaction is created **and cancel is set to false** so we can save again an item and commit it with its previous and or new mutations.

**Maybe yesSql would have to handle some cases**, maybe after a `Cancel()` never reset cancel to false even if we create a new tansaction to query again items. After a `CommitAsync()` i understand that we may want separate transactions, but at least maybe clear the identityMap to not reuse an entity already mutated in a previous canceled transaction.

**So about session helper i will think about it this w-end and will let you know ;)**

Note: And i don't talk about a WF triggered by a content handler so acting as an handler (if no intermediate waiting activity defers the execution) and that uses a `RegisterUserTask` that creates an user through the `UserStore` that does a `CommitAsync` (as it has to return an `IdentityResult` immediately) that would short circuit e.g. a controller that would have done a session cancel.

**Oups i talk too much, sorry to bother your PR, but again nothing to block it**, time to sleep ;)

deanmarcussen commented:
cool @jtkech 
> But i also saw that if we do a mutation + save, then if we do a session cancel but then another query + save the mutation is persisted. So it is important that a cancel is the last operation before committing, as in controllers. **So here the session helper re-doing the save is not good**, we would need to do all cancel through the session helper so that it could check it before re-doing a save.

Ah this is interesting. 
Ok I checked here, to make sure that cancel is the last operation performed, before an exception, so here it is good.

So I will leave this as ready to merge for when @Skrypt has time to finish his review too ;)
